src/net/netsyscall.c

```diff
@@ -1702,6 +1702,11 @@ void register_net_syscalls(struct syscall *map)
     register_syscall(map, shutdown, shutdown);
 }
 
+closure_function(0, 0, void, netsock_poll) {
+    netif_poll_all();
```

Comment:
I know the comment above netif_poll states that it should be called in a main loop, but we may want to question if this is absolute necessary, as it means running this routine every time through runloop (when kernel lock is acquired). (And it's really a flaw in runloop that the CPU can go idle while something is sitting on a runqueue...if that was corrected this would run endlessly and never sleep.)

Is it only needed for loopback function and, if so, could it be triggered by net syscalls instead? Or perhaps run from a periodic timer if the loopback interface is actively in use (has buffers queued)?

Reply:
I agree it's better not to rely on the fact that the scheduler doesn't immediately serve re-queued items in the runqueue. Yes, calling netif_poll() is only needed for loopback function, and as such it can be triggered by anything that sends a network packet. How about the updated code, where the thunk is put in the runqueue only when data is sent on a socket (and in case of TCP when a connection/disconnection is done)?